modules/sdk-lib-mpc/src/tss/ecdsa-dkls/dsg.ts

```diff
@@ -0,0 +1,169 @@
+import { SignSession, Keyshare, Message } from '@silencelaboratories/dkls-wasm-ll-node';
+import { DeserializedBroadcastMessage, DeserializedMessages, DklsSignature, DsgState } from './types';
+import { decode } from 'cbor';
+
+export class Dsg {
+  protected dsgSession: SignSession | undefined;
+  protected dsgSessionBytes: Uint8Array;
+  protected signature: DklsSignature | undefined;
+  protected keyShare: Keyshare;
+  protected messageHash: Buffer;
+  protected derivationPath: string;
+  protected partyIdx: number;
+  protected dsgState: DsgState = DsgState.Uninitialized;
+
+  constructor(keyShare: Buffer, partyIdx: number, derivationPath: string, messageHash: Buffer) {
+    this.keyShare = Keyshare.fromBytes(keyShare);
+    this.partyIdx = partyIdx;
+    if (this.keyShare.partyId !== partyIdx) {
+      throw Error(`Party index: ${partyIdx} does not match key share partyId: ${this.keyShare.partyId} `);
+    }
+    this.derivationPath = derivationPath;
+    this.messageHash = messageHash;
+  }
+
+  private _restoreSession() {
+    if (!this.dsgSession) {
+      this.dsgSession = SignSession.fromBytes(this.dsgSessionBytes);
+    }
+  }
+
+  private _deserializeState() {
+    if (!this.dsgSession) {
+      throw Error('Session not intialized');
+    }
+    const round = decode(this.dsgSession.toBytes()).round;
+    switch (round) {
+      case 'WaitMsg1':
+        this.dsgState = DsgState.Round1;
+        break;
+      case 'WaitMsg2':
+        this.dsgState = DsgState.Round2;
+        break;
+      case 'WaitMsg3':
+        this.dsgState = DsgState.Round3;
+        break;
+      case 'Ended':
+        this.dsgState = DsgState.Complete;
+        break;
+      default:
+        this.dsgState = DsgState.InvalidState;
+        throw `Invalid State: ${round}`;
+    }
+  }
+
+  async initDsg(): Promise<DeserializedBroadcastMessage> {
+    if (this.dsgState !== DsgState.Uninitialized) {
+      throw 'DSG session already initialized';
+    }
+    if (typeof window !== 'undefined') {
+      const initDkls = require('@silencelaboratories/dkls-wasm-ll-web');
+      await initDkls();
+    }
+    this.dsgSession = new SignSession(this.keyShare, this.derivationPath);
+    try {
+      const payload = this.dsgSession.createFirstMessage().payload;
+      this._deserializeState();
+      return {
+        payload: payload,
+        from: this.partyIdx,
+      };
+    } catch (e) {
+      throw `Error while creating the first message from party ${this.partyIdx}: ${e}`;
+    }
+  }
+
+  getSignature(): DklsSignature {
```

Comment:
https://www.typescripttutorial.net/typescript-tutorial/typescript-getters-setters/
```suggestion
  get signature(): DklsSignature {
```

Reply:
done